### PR TITLE
Sankey with Recharts

### PR DESCRIPTION
**Pull request by shaankhosla**
This PR introduces a new plot, a Sankey diagram of income and expenses,  created with a new charting library Recharts.

[Recharts](https://recharts.org/en-US) is an open-source React chart library with ~21k stars and seemed to be the easiest to use out of the other libraries I looked at. I think this Sankey diagram, and Recharts, would be a great addition to the app. IMO, the Sankey diagram looks great with the very little customization that I did, but I'm excited for others to try it with their own data. This PR is a WIP for 2 reasons:

1. This is dependent on first proving that Recharts would fit with Actual, and therefore I'll try to create another PR soon that ports over the Networth report to Recharts. This will help us decide if this is the right library to move to.

2. I'm having some trouble getting the chart to render correctly on the Report Overview page. I'm sure this can be fixed relatively easily but I haven't spent much time on it yet.

### CODE REVIEW

**Comment by shaankhosla** [comment]
Networth graph is recreated with Recharts in #1740

File: package.json [hunk 1]
```diff
@@ -64,5 +64,8 @@
   "browserslist": [
     "electron 24.0",
     "defaults"
-  ]
+  ],
+  "dependencies": {
+    "recharts": "^2.8.0"
```
Comment (joel-jeremy) [inline package.json:69]: We should add this to the `desktop-client` package instead of the root package.

**Comment by joel-jeremy** [comment]
Suggestion: Is it possible to allow zooming in and out of the charts? It gets to a certain points where hovering parts of the chart becomes a challenge because they get so small.

**Comment by shaankhosla** [comment]
> Suggestion: Is it possible to allow zooming in and out of the charts? It gets to a certain points where hovering parts of the chart becomes a challenge because they get so small.

I don't zooming in and out is supported, but we could regroup all small categories towards an "Other" category. I'm not sure exactly what the threshold would be, but it should work fine.

File: package.json (hunk 1 above)
Comment (shaankhosla) [reply package.json]: Done :)

**Comment by shaankhosla** [comment]
> Not sure how the charts handle decimal values but it seems to handle them as floating points. Should we do dome truncation here? image

I've added truncation to the tooltip

**Comment by nmathey** [comment]
This is kind of personal customization/wish but Is it possible to calculate and display % at this stage (in addition or in place of the amount) ?

**Comment by shaankhosla** [comment]
> Suggestion: This is kind of personal customization/wish but Is it possible to calculate and display % at this stage (in addition or in place of the amount) ?

Yeah that'd be neat, I'd like that as well. Might be a little more complicated so unsure if that should be included in this PR. Especially since we're not even 100% sure this will be merged in the first place.

**Comment by Crazypkr1099** [comment]
Great proof of concept but I think it needs alot more work to be merged. The data is not easy to be seen for one issue, which I suggest having the name of each line visible. Secondly, colors are also needed to differentiate between each line. 

Something like this: https://codesandbox.io/s/8b3hm?file=/src/index.js (I didn't create this, found it on the web)

Looks great though so far. Can't wait to see it finished.

**Comment by shaankhosla** [comment]
> Great proof of concept but I think it needs alot more work to be merged. The data is not easy to be seen for one issue, which I suggest having the name of each line visible. Secondly, colors are also needed to differentiate between each line.
> 
> Something like this: https://codesandbox.io/s/8b3hm?file=/src/index.js (I didn't create this, found it on the web)
> 
> Looks great though so far. Can't wait to see it finished.

That was a helpful link, I've added the labels to the nodes. 

image

**Comment by MatissJanis** [comment]
👋 Exciting! Sadly I'm unable to test it with my budget as there seems to be a bug somewhere. But here are my notes from using the limited demo budget:

```
TypeError: undefined is not an object (evaluating 't.data[0].name')
```

- [ ] homepage: there is no "summary" graph available; I'm also unsure what would best fit there for sankey.. Maybe a minimal sankey with only income (all income categories grouped in 1x); "starting budget" and outcome (again everything grouped in a single category)?
- [ ] full page: the labels shouldn't be bold
- [ ] full page: can we add some colors? The spending report has some color logic - would be nice to re-use that
- [ ] homepage: when both "category spending" and "sankey reports" features are enabled the homepage has a weird layout
- [ ] full page: as @MikesGlitch reported before - perhaps the easiest would be to disable the animation for tooltips? I do agree the current "flying" effect is quite off-putting

Let me know when you want me to do a full code review.

**Comment by nmathey** [comment]
> homepage: there is no "summary" graph available; I'm also unsure what would best fit there for sankey.. Maybe a minimal sankey with only income (all income categories grouped in 1x); "starting budget" and outcome (again everything grouped in a single category)?

Or maybe a mix of your two proposition :) --> One unique "Income" group and "groups of categogies" as outcomes.
I guess it depend on how everyone is building its budget. On my side I'm following the [50/30/20 rule](https://www.unfcu.org/guides/the-50-30-20-rule/#:~:text=Do%20not%20subtract%20other%20amounts,%2C%20and%2020%25%20toward%20savings.) so I will have basically have "Needs", "Savings" and "Wants" groups of categories.

**Comment by shaankhosla** [comment]
> 👋 Exciting! Sadly I'm unable to test it with my budget as there seems to be a bug somewhere. But here are my notes from using the limited demo budget:
> 
> ```
> TypeError: undefined is not an object (evaluating 't.data[0].name')
> ```
> 
>     * [ ]  homepage: there is no "summary" graph available; I'm also unsure what would best fit there for sankey.. Maybe a minimal sankey with only income (all income categories grouped in 1x); "starting budget" and outcome (again everything grouped in a single category)?
> 
>     * [ ]  full page: the labels shouldn't be bold
> 
>     * [ ]  full page: can we add some colors? The spending report has some color logic - would be nice to re-use that
> 
>     * [ ]  homepage: when both "category spending" and "sankey reports" features are enabled the homepage has a weird layout
> 
>     * [ ]  full page: as @MikesGlitch reported before - perhaps the easiest would be to disable the animation for tooltips? I do agree the current "flying" effect is quite off-putting
> 
> 
> Let me know when you want me to do a full code review.

Hi @MatissJanis, probably not ready for a full-code review but I've made a bit of progress on these points. 

1. Try your budget again, I believe the bug is fixed
2. I'll hold off on the overview graph since I'm unsure what this should look like at the moment. 
3. Labels have been unbolded
4. Homepage layout with the "category spending" chart has been fixed.
5. Tooltip animation has been disabled.

I also don't think the color logic from the category spending would look good on this report, it'll probably look too crowded and noisy

**Comment by MatissJanis** [comment]
👋 Apologies for the slow review. It's been a very busy time for me.

This is looking much better since I last reviewed! Here's some more small pieces of feedback:

- [ ] homepage: could we either a) add a left-right padding to the graph or b) remove the bottom padding? Currently it looks inconsistent when the spending graph is enabled (screenshot)
- [ ] nitpick: can we remove the extra space between the ":" and the category? i.e. "Something - here : something" vs "Something - here: something" (screenshot)

Screenshot 2023-10-22 at 16 50 42
Screenshot 2023-10-22 at 16 51 08

File: packages/desktop-client/src/components/reports/graphs/sankey-spreadsheet.tsx [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      let payeeNames = {};
+      for (let id in payeesDict) {
+        const result = await runQuery(
+          q('payees').filter({ id: id }).select(['name']),
+        );
+        if (result.data.length > 0 && result.data[0].name) {
+          payeeNames[result.data[0].name] = payeesDict[id];
+        }
+      }
+      return payeeNames;
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/graphs/sankey-spreadsheet.tsx:111]: **💬 suggestion:** ‏could we make a single query here with multiple IDs and then convert the resulting array to a payee-name-map that you need? That would be much more optimal.

**Comment by shaankhosla** [comment]
@MatissJanis thanks for your feedback! I've fixed the edits you've suggested. I'm unsure why it's still failing with your budget.

**Comment by MatissJanis** [comment]
So I debugged a bit and I think the problem is that in `data.nodes` I have some duplicate names. I'm guessing you're missing some check somewhere to make the names unique?

**Comment by shaankhosla** [comment]
> So I debugged a bit and I think the problem is that in `data.nodes` I have some duplicate names. I'm guessing you're missing some check somewhere to make the names unique?

Hm, I'm not sure how - I check to make sure there aren't duplicate nodes with `nodeNames` in `packages/desktop-client/src/components/reports/graphs/sankey-spreadsheet.tsx`

**Comment by MatissJanis** [comment]
Nice, so it works for my budget now. But.. I wonder if we could improve it further for real budgets?

For example: instead of showing all income categories - show only the top ones and then group the rest in "other" (I currently have 2 bigger income categories and then a bunch of tiny ones).

Another cool improvement would be: when you click on the category: filter to show only that. So for example - if I click on a top level category - it removes everything on the left side and only shows the clicked category + children.

What do you think?

Screenshot 2023-10-23 at 21 49 17

**Comment by shaankhosla** [comment]
That makes sense, I've been trying to implement the following:
1. Make an "Other" category at each level (Income, category, and subcategories) that merges nodes with small values (maybe using some threshold like 1/16th of the largest node on the graph). I have a lot of different subcategories with small values in them and it makes my graph look a little odd too
2. Clicking on a category, and then filtering the graph to only show the clicked category and it's respective subcategories, is a great idea and should definitely be added.

I've been unable to implement either of these features so far. Would you be open to merging what's here, under a feature flag, which would let others run with this? Or perhaps I could remove the subcategories (so that it's just Income payee nodes -> Budget -> Category nodes) to make it a bit cleaner visually?

**Comment by youngcw** [comment]
How would this fit in with the up coming custom reports?  Would it be possible to modify this PR to be a report type in the custom reports?

**Comment by shaankhosla** [comment]
I'd like to see this get merged in as is as a default dashboard visualization (currently behind a feature flag). Combining it with the custom reports might be challenging since there's not many ways you can modify this Sankey plot using the options available on custom reports, and in many of those different configurations I don't think this plot would even render correctly. 

Are we planning on removing all the current default dashboard visualizations since we'll have custom reports soon? If not, I'd like to see this become one of those.

**Comment by youngcw** [comment]
I don't know of any plans to remove the other reports, but I think it could make sense after custom reports get finished and moved out of experimental.  Thats just my thoughts though. For the time being it probably makes more sense to get this merged in under a flag.

**Comment by shaankhosla** [comment]
Yeah makes sense, I think it'd be useful to have some visualizations available out of the box so it might be nice to keep some standard visualizations on the dashboard.

Since this sankey chart is already behind a flag, is there anything else you'd need to see added before this is merged in?

File: packages/desktop-client/src/components/reports/graphs/sankey-spreadsheet.tsx [hunk 3]
```diff
@@ -0,0 +1,191 @@
+import q, { runQuery } from 'loot-core/src/client/query-helpers';
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/graphs/sankey-spreadsheet.tsx:1]: **👏 praise:** ‏thanks for creating this file with TS

File: packages/desktop-client/src/components/reports/graphs/sankey-spreadsheet.tsx [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+      conditions: conditions.filter(cond => !cond.customName),
+    });
+    const conditionsOpKey = conditionsOp === 'or' ? '$or' : '$and';
+
+    // create list of Income subcategories
+    const allIncomeSubcategories = [].concat(
+      ...categories
+        .filter(category => category.is_income === 1)
+        .map(category => category.categories),
+    );
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/graphs/sankey-spreadsheet.tsx:24]: **💬 suggestion:** ‏unnecessary array spread + concat.

```suggestion
    // create list of Income subcategories
    const allIncomeSubcategories = categories
        .filter(category => category.is_income === 1)
        .map(category => category.categories);
```

File: packages/desktop-client/src/components/reports/Overview.js [hunk 5]
```diff
@@ -325,12 +327,48 @@ function CategorySpendingCard() {
...
+}
+
 export default function Overview() {
   let categorySpendingReportFeatureFlag = useFeatureFlag(
     'categorySpendingReport',
   );
+  let sankeyFeatureFlag = useFeatureFlag('sankeyReport');
 
   let accounts = useSelector(state => state.queries.accounts);
+  let categories = useSelector(state => state.queries.categories.grouped);
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/Overview.js:371]: **💬 suggestion:** ‏you might want to use `useCategories` here. It fetches the categories if they have not been fetched already. Whereas your current implementation just returns the cache.

File: packages/desktop-client/src/components/reports/Sankey.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+import View from '../common/View';
+
+import sankeySpreadsheet from './graphs/sankey-spreadsheet';
+import SankeyGraph from './graphs/SankeyGraph';
+import Header from './Header';
+import useReport from './useReport';
+import { fromDateRepr } from './util';
+
+export default function Sankey() {
+  let categories = useSelector(state => state.queries.categories.grouped);
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/Sankey.js:21]: Same note here

**Review by MatissJanis — APPROVED** [review]
Overall LGTM! My (big) budget still looks a bit wonky.. but we can iterate after merging since this is under a feature flag. :+1:

Thanks for the hard work here!

**Pull request merged**